pacta/email.go

```diff
@@ -0,0 +1,53 @@
+package pacta
+
+import (
+	"fmt"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+
+	normalizer "github.com/dimuska139/go-email-normalizer"
+)
+
+var n = normalizer.NewNormalizer()
+
+func CanonizeEmail(email string) (string, error) {
+	// Performs basic validation, preventing obviously malformed email addresses, but
+	// not capturing the entirety of the RFC 5322 grammar.
+	split := strings.Split(email, "@")
+	if len(split) != 2 {
+		return "", fmt.Errorf("invalid email, wrong number of at-signs: %q", email)
+	}
+	if !isASCII(split[0]) || !utf8.ValidString(split[0]) {
+		return "", fmt.Errorf("invalid email, non-ASCII or UTF-8 local part: %q", split[0])
+	}
+	if includesSpace(email) {
+		return "", fmt.Errorf("invalid email, whitespace in email address: %q", split[0])
+	}
+	if len(split[0]) == 0 || len(split[1]) == 0 {
+		return "", fmt.Errorf("invalid email, empty local or domain part: %q", email)
+	}
+	// Performs complex, vendor-specific normalizations, preventing obvious duplicates
+	// like plus-aliases, and gmail dot aliases. This will return the original if it cannot
+	// parse or simplify it.
+	return n.Normalize(email), nil
+}
+
+func isASCII(s string) bool {
+	for _, r := range s {
+		if r > 127 {
+			return false
+		}
+	}
+	fmt.Printf("%s = true\n", s)
```

Comment:
Remove, though per the previous comment this whole function should prob go

Reply:
GC on removing this line ;)